Adding to the contrib module an Proxy Actor implementing the circuit breaker pattern

galarragas opened the pull request:
In some of my project I needed to use a circuit breaker for actor request-reply communication, mostly in Spray application for the connection from the REST API layer to the service actors. 
I re-implemented the same state machine documented in the AKKA CircuitBreaker pattern as an actor proxy with some small adaptation required by the different underlying communication mechanism. 

It worked quite well for me so I think it might be useful as a shared resource.

galarragas committed:
Adding to the contrib module an Proxy Actor implementing the circuit breaker pattern

typesafehub-validator commented:
Hi @galarragas,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

patriknw commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+
+This is an alternative implementation of the [AKKA Circuit Breaker Pattern](http://doc.akka.io/docs/akka/snapshot/common/circuitbreaker.html).
+The main difference is that is intended to be used only for request-reply interactions with actor using the Circuit-Breaker as a proxy of the target one
+in order to provide the same failfast functionnalities and a protocol similar to the AKKA Pattern implementation
```
functionalities

patriknw commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,266 @@
+  sealed trait CircuitBreakerEvent
+  case class CircuitOpen(circuit: ActorRef) extends CircuitBreakerCommand
+  case class CircuitClosed(circuit: ActorRef) extends CircuitBreakerCommand
+  case class CircuitHalfOpen(circuit: ActorRef) extends CircuitBreakerCommand
```
make the case classes `final`

patriknw commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,266 @@
+
+    target.ask(message)(callTimeout).onComplete {
+      case Success(response) ⇒
+        log.debug("Request '{}' has been replied to with response {}, forwarding to original sender {}", message, currentSender)
```
to be strict, the `log` from ActorLogging should only be accessed (initialized) by the actor. I don't think we will see any problems from this but more correct would be to pass the `LoggingAdapter` as a parameter to this method

galarragas replied:
Hi @patriknw : I applied all the changes, very good point on the log, I completely overlooked it..

patriknw commented:
great contribution, @galarragas
LGTM

patriknw commented:
OK TO TEST

galarragas committed:
Post revision changes:

- Typo in documentation
- Made case classes `final`
- Hidden the actor's `log` with one passed as parameter to the method

patriknw commented:
thanks, LGTM

ktoso commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+Circuit-Breaker Actor
+======================
+
+This is an alternative implementation of the [AKKA Circuit Breaker Pattern](http://doc.akka.io/docs/akka/snapshot/common/circuitbreaker.html).
```
s/AKKA/Akka/ :-)

ktoso commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+Circuit-Breaker Actor
+======================
```
this will cause a build warning, please align the underscore:

```
Circuit-Breaker Actor
=====================
```

ktoso commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+======================
+
+This is an alternative implementation of the [AKKA Circuit Breaker Pattern](http://doc.akka.io/docs/akka/snapshot/common/circuitbreaker.html).
+The main difference is that is intended to be used only for request-reply interactions with actor using the Circuit-Breaker as a proxy of the target one
```
The main difference is that *it* is indented

ktoso replied:
with *an* actor

ktoso commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+
+This is an alternative implementation of the [AKKA Circuit Breaker Pattern](http://doc.akka.io/docs/akka/snapshot/common/circuitbreaker.html).
+The main difference is that is intended to be used only for request-reply interactions with actor using the Circuit-Breaker as a proxy of the target one
+in order to provide the same failfast functionalities and a protocol similar to the AKKA Pattern implementation
```
similar to the circuit-breaker implementation in Akka**.**

ktoso commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+         }
+
+     }
+   }
```
This is a lot of code (below snippets too) to be put literally in the docs. 
Please create a file in tests which contains these, and then use `.. includecode::`
(e.g. `.. includecode:: code/docs/stream/io/StreamFileDocSpec.scala#file-source`)

ktoso commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+     }
+   }
+
+Using the ASK pattern, in this case it is useful to be able to map circuit open failures to the same type of failures
```
Not caps please, let's call it:

```
Using the ``ask`` pattern

```

ktoso commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+   }
+
+Using the ASK pattern, in this case it is useful to be able to map circuit open failures to the same type of failures
+returned by the service (a `Left[String]` in our case): ::
```
in restructured text, this must be:

```
e (a ``Left[String]`` in our case): ::
```

ktoso commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+
+
+If it is not possible to define define a specific error response, you can map the Open Circuit notification into a failure.
+That also means that your `CircuitBreakerActor` will be essentially useful to protect you from time out for extra workload or
```
restructured text format is for inline code is:

```
``CircuitBreakerActor`` w
```

http://docutils.sourceforge.net/docs/user/rst/quickstart.html

ktoso replied:
`will be essentially useful to protect` please remove essentially - just useful is enough

ktoso commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+   }
+
+
+If it is not possible to define define a specific error response, you can map the Open Circuit notification into a failure.
```
to a failure

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,266 @@
+   * Creates an circuit breaker actor proxying a target actor intended for request-reply interactions.
+   * It is possible to send messages through this proxy without expecting a response wrapping them into a
+   * [[akka.contrib.circuitbreaker.CircuitBreakerActor.TellOnly]]
+   *
```
:+1: good docs here

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,266 @@
+   *                         response from a request represent a failure
+   * @param failureMap   function to map a failure into a response message. The failing response message is wrapped
+   *                     into a [[akka.contrib.circuitbreaker.CircuitBreakerActor.CircuitOpenFailure]] object
+   * @return
```
please remove empty `@return` line

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,266 @@
+     * Creates the props for a [[akka.contrib.circuitbreaker.CircuitBreakerActor]] proxying the given target
+     *
+     * @param target the target actor ref
+     * @return
```
please remove

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,266 @@
+    /**
+     * Creates the props for a [[akka.contrib.circuitbreaker.CircuitBreakerActor]] proxying the given target
+     *
+     * @param target the target actor ref
```
please remove, does not add much information

ktoso replied:
ping on this one :)

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,266 @@
+   *   [[akka.contrib.circuitbreaker.CircuitBreakerActor.CircuitOpenFailure]] failure responses throwing
+   *   an exception built with the given exception builder
+   *
+   * @param future
```
please remove, empty

ktoso commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,184 @@
+### Usage
+
+Let's assume we have an actor wrapping a back-end service and able to respond to `Request` calls with a `Response` object
+containing an `Either[String, String]` to map successful and failed responses. The service is also potentially slowing down
```
```
``Either[String, String]``
```

ktoso commented:
Went trough docs and will go over impl soon - need to head out for now though.
I'll continue the review soon :)

galarragas committed:
Post revision changes:

- Moved sample code in source files
- Text review
- Proper use of RST syntax
- Improved ScalaDoc

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,247 @@
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success }
+
+object CircuitBreakerActor {
```
Perhaps we should also add a doc here that "This is an Actor which implements the circuit breaker pattern, you may also be interested in the raw circuit breaker [[akka.pattern.CircuitBreaker]]"?

ktoso replied:
ping here :)

galarragas replied:
:+1:

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,247 @@
+import CircuitBreakerActor._
+import CircuitBreakerInternalEvents._
+
+class CircuitBreakerActor(
```
`final`?

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,247 @@
+  implicit class CircuitBreakerAwareFuture(val future: Future[Any]) extends AnyVal {
+    def failForOpenCircuit(implicit executionContext: ExecutionContext): Future[Any] = failForOpenCircuitWith(new OpenCircuitException)
+
+    def failForOpenCircuitWith(throwing: ⇒ Throwable)(implicit executionContext: ExecutionContext): Future[Any] = {
```
`executionContext` not needed, we eagerly created failed/successful

ktoso commented on akka-contrib/src/test/scala/akka/contrib/circuitbreaker/CircuitBreakerActorSpec.scala:
```diff
@@ -0,0 +1,302 @@
+      sender.expectMsg(reply)
+    }
+
+    def waitForCircuitBreakerToReceiveSelfNotificationMessage = Thread.sleep(baseCircuitBreakerBuilder.resetTimeout.duration.toMillis / 4)
```
expectNoMsg instead of sleep explicitly?

ktoso commented on akka-contrib/src/test/scala/akka/contrib/circuitbreaker/sample/CircuitBreakerAsk.scala:
```diff
@@ -0,0 +1,54 @@
+package akka.contrib.circuitbreaker.sample
```
Please add license headers to all new files:

```
/**
 * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
 */
```

Thanks!

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,247 @@
+  }
+
+  when(Closed) {
+    callSucceededHandling orElse {
```
`CallSucceeded` seems unlikely to arrive when in `Closed` state as I read it. Shall we re-order here, such that the most frequent message is first (the partial function, then `orElse callSucceededHandling`)?

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,247 @@
+
+  final case class CircuitBreakerStateData(failureCount: Int = 0, firstHalfOpenMessageSent: Boolean = false)
+
+  final case class CircuitBreakerActorBuilder(
```
Not sold on `Builder` in the name.
Also, since it's a proxy.. maybe `CircuitBreakerProxy`?

ktoso replied:
Needs to get Java API, so all the methods need to gain `withMaxFailures(Int)`, `withFailureDetector(...)`, etc.

galarragas replied:
Hi @ktoso what about `CircuitBreakerPropsBuilder` ? The only purpose of this class is to help in building the `props` object for the `CircuitBreakerActor`...

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,247 @@
+
+  }
+
+  class OpenCircuitException extends Exception("Circuit Open so unable to complete operation")
```
final

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,247 @@
+     *
+     * @param target the target actor ref
+     */
+    def propsForTarget(target: ActorRef) = CircuitBreakerActor.props(target, maxFailures, callTimeout, resetTimeout, circuitEventListener, failureDetector, openCircuitFailureConverter)
```
just `props()` should be enough I think (one can call it like `props(target = ref)` if one wants to)

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerActor.scala:
```diff
@@ -0,0 +1,247 @@
+        stay
+
+      case Event(openNotification @ CircuitOpenFailure(_), _) ⇒
+        log.error("Why did I send message {} to myself?", openNotification)
```
If this happens its a bug in this impl - let's explain that, so *if* it occurs users are not completely confused by this error message. (We can ask them to report it, or we should remove this case completely)

galarragas replied:
Actually that code has been introduced when I was trying to fix a bug :-) I'll improve the logging with a proper message...

ktoso commented on akka-contrib/src/test/scala/akka/contrib/circuitbreaker/CircuitBreakerActorSpec.scala:
```diff
@@ -0,0 +1,302 @@
+      Then("An event is sent")
+      eventListener.expectMsg(CircuitClosed(circuitBreaker))
+
+    }
```
needs a test (and impl, since it does not yet do that) for this scenario:
- circuit breaker forwards a message which causes the actor to stop
- circuit breaker should stop as well, since it'll now just keep sending into deadLetters otherwise.

This can be solved by doing a `context.watch(target)` in the `preStart` of the CircuitBreakerProxy actor.

galarragas replied:
Nice, I'll work on that

ktoso commented on akka-contrib/src/test/scala/akka/contrib/circuitbreaker/sample/CircuitBreakerAskWithFailure.scala:
```diff
@@ -0,0 +1,38 @@
+
+  override def receive: Receive = {
+    case AskFor(requestToForward) ⇒
+      (serviceCircuitBreaker ? Request(requestToForward)).failForOpenCircuit.mapTo[String].onComplete {
```
I'm not completely convienved about `failForOpenCircuit` yet (adding methods to Future via an implicit via importing `import akka.contrib.circuitbreaker.CircuitBreakerActor._` seems hard to discover). I see why it's useful - to keep the fluent-api going to the mapTo etc, how we obtain this method is hard to discover

galarragas replied:
I'm going to try a couple of things. One is to extract the `failForOpenCircuit` enabling implicit in a different object since I agree with you that current way is a bit obscure.
The second thing I was thinking about was to have a method `withCircuitBreaker` (or `failForOpenCircuit` for consistency) to use around the future with this syntax:

```(scala)
withCircuitBreaker {  serviceCircuitBreaker ? Request(requestToForward) }.mapTo[String].onComplete {
```

ktoso commented on akka-contrib/src/test/scala/akka/contrib/circuitbreaker/sample/SimpleService.scala:
```diff
@@ -0,0 +1,45 @@
+
+  }
+}
+//#simple-service
```
Maybe put the 3 examples into 1 file, they logically belong together and that would read easier.

ktoso commented on akka-contrib/src/test/scala/akka/contrib/circuitbreaker/sample/package.scala:
```diff
@@ -0,0 +1,5 @@
+
+package object sample {
+  case class AskFor(what: String)
+}
```
please avoid package objects.
Let's put this in the Service actor's companion.

In docs it should also be explicit that this is the example Service's API, and not part of the circuit breaker API.

galarragas committed:
Post revision changes:

- Added reference to the circuit breaker pattern function in the CircuitBreakerActor scaladoc
- made `final` classes `OpenCircuitException` and `CircuitBreakerActor`
- Renamed `CircuitBreakerActorBuilder` into `CircuitBreakerActorPropsBuilder` and renamed the `propsForTarget` method into `props`

galarragas committed:
Post revision changes:

- Renamed the circuit breaker actor into a proxy for clarity and small renaming of the builder
- Added Java-friendly `with` copy methods to `CircuitBreakerPropsBuilder`
- Refactored a test not using the scenario trait as others

galarragas committed:
Post revision changes:

- Reordered message cases for Closed state management
- Improved error message for unexpected self-sent circuit open message

galarragas committed:
Post revision changes:

- Added handling of termination of target actor
- Introduced a Passthrough command type to allow sending message to the target actor in any state. It is intended for control messages such as `PoisonPill` or `Kill`

galarragas committed:
Post revision changes:

- Refactored the part associated to the `failForOpenCircuit` extension to Future was done.
  - Moved implicit conversion into a specific class. Not requiring anymore the `._` import but importing only the implicit conversion needed
  - Added a `askWithCircuitBreaker` approach to replace the ask and the following `failForOpenCircuit`
- Compacted all samples in the same file and added the copyright header
- Improved documentation

galarragas committed:
Trying to fix build failures about not thrown exceptions in ScalaDoc and not found reference to `akka.pattern.ask`

galarragas committed:
Trying to fix build failures about not thrown exceptions in ScalaDoc

galarragas commented:
Sorry for the many commits. I didn't squash my commits, I thought it would have been easier to link the changes to the comment this way. I also haven't been able to replicate locally the ScalaDoc validations so I had to run them on the CI

galarragas committed:
Added copyright header to `askExtensions.scala` file and did some small improvement to ScalaDocs and rst documentation

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/askExtensions.scala:
```diff
@@ -0,0 +1,94 @@
+ * `Future` result of an `ask` pattern to fail in case of
+ * [[akka.contrib.circuitbreaker.CircuitBreakerProxy.CircuitOpenFailure]] response
+ */
+object implicits {
```
upper case - like we have in other places.

ktoso commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/askExtensions.scala:
```diff
@@ -0,0 +1,94 @@
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+final class OpenCircuitException extends Exception("Unable to complete operation since the Circuit Breaker Actor Proxy is in Open State")
```
`RuntimeException` - because otherwise it's a checked one for JavaDSLs, not as nice to work with

galarragas committed:
Upper case for `Implicits` object and made the CircuitOpenException extending RuntimeException

ktoso commented:
Oh, PR was updated - will review again then. Thanks @galarragas – please ping if you update a PR, or we won't notice as quickly :)

galarragas commented:
Hi @ktoso I will do. Sorry I forgot about it

ktoso commented:
No problem :) Will get around to reviewing soon

johanandren commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerProxy.scala:
```diff
@@ -0,0 +1,258 @@
+  }
+}
+
+object CircuitBreakerInternalEvents {
```
Private to CircuitBreakerProxy inside the companion object, since they are internal?

johanandren commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerProxy.scala:
```diff
@@ -0,0 +1,258 @@
+        stay
+
+      case Event(openNotification @ CircuitOpenFailure(_), _) ⇒
+        log.error("Unexpected circuit open notification {} sent to myself. Please report this as a bug.", openNotification)
```
It would be a bug, but not really an error, nothing breaks, right?

galarragas replied:
Correct, moving the log to `warning` level

johanandren commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerProxy.scala:
```diff
@@ -0,0 +1,258 @@
+  when(Closed) {
+    commonStateHandling orElse {
+      case Event(TellOnly(message), _) ⇒
+        log.debug("CLOSED: Sending message {} without expecting any response", message)
```
Inconsistency-nitpick: the state is called Closed but log says CLOSED, same for debug logs below.

johanandren commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,54 @@
+envelope. The difference between the two is that ``TellOnly`` will forward the message only when in closed mode and
+``Passthrough`` will do it in any state. You can for example use the ``Passthrough`` envelope to wrap a ``PoisonPill``
+message to terminate the target actor. That will cause the circuit breaker proxy to be terminated too
+
```
Nice and thorough docs! :+1:

johanandren commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/askExtensions.scala:
```diff
@@ -0,0 +1,100 @@
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+final class OpenCircuitException extends RuntimeException("Unable to complete operation since the Circuit Breaker Actor Proxy is in Open State")
```
Could be an object since it will contain the same message all the time?

galarragas replied:
I wasn't sure how the Java compatibility of that would have been. I am thinking to create a  `OpenCircuitException` as a single instance extending the class `OpenCircuitException` to be used by the Java code. Would that be good? or is it useless?

johanandren commented on akka-contrib/src/main/scala/akka/contrib/circuitbreaker/askExtensions.scala:
```diff
@@ -0,0 +1,100 @@
+ * [[akka.contrib.circuitbreaker.CircuitBreakerProxy.CircuitOpenFailure]] failure responses throwing
+ * an exception built with the given exception builder
+ */
+class CircuitBreakerAwareFuture(val future: Future[Any]) extends AnyVal {
```
Final?

johanandren commented on akka-contrib/src/test/scala/akka/contrib/circuitbreaker/CircuitBreakerProxySpec.scala:
```diff
@@ -0,0 +1,536 @@
+    }
+
+    "enter open state after reaching the threshold of timed-out responses" in new CircuitBreakerScenario {
+      Given("A circuit breaker actor proxying a test probe")
```
Not that I really mind but I haven't seen the given-when-then style anywhere else in the akka code base. Shouldn't it at least be used consistently in all test cases in this test if used?

johanandren commented on akka-contrib/src/test/scala/akka/contrib/circuitbreaker/CircuitBreakerProxySpec.scala:
```diff
@@ -0,0 +1,536 @@
+      Then("The message is sent to the target actor")
+      receiver.expectMsg("request")
+
+      When("Then target actor replies")
```
Isn't the whole idea with given-when-then that that is the entire list of states, here it repeats.

johanandren commented on akka-contrib/src/test/scala/akka/contrib/circuitbreaker/CircuitBreakerProxySpec.scala:
```diff
@@ -0,0 +1,536 @@
+    }
+  }
+
+}
```
In general very nice and clear tests!

johanandren commented:
LGTM with some nitpicking (and some openish:questions which might not require any further changes)

galarragas committed:
Post review changes:
- Moved `CircuitBreakerInternalEvents` inside `CircuitBreakerProxy` companion object,
- changed logs messages to be consistent with the state name case,
- used the correct log level for the case of unexpected self-sending of `CircuitOpenFailure`
- Converted the  `OpenCircuitException` class into an object implementing the sealed class `OpenCircuitException`
- `CircuitBreakerAwareFuture` is a final class
- Using Given-When-Then consistently across all tests (can remove them all if we prefer to keep consistency with the rest of the project)
- Using Given-When-Then consistently across all tests (can remove them all if we prefer to keep consistency with the rest of the project)

galarragas commented:
Hi @johanandren . I applied the suggestions you did during the review. Let me know if you have any other comments. Thanks

johanandren commented:
Nope, LGTM now, thanks a lot!

rkuhn commented on akka-contrib/docs/circuitbreaker.rst:
```diff
@@ -0,0 +1,54 @@
+Circuit-Breaker Actor
+=====================
+
+This is an alternative implementation of the [Akka Circuit Breaker Pattern](http://doc.akka.io/docs/akka/snapshot/common/circuitbreaker.html).
```
This should use a sphinx ref.

galarragas committed:
Post review changes: converted reference to `circuit-breaker` into a rst ref. Renamed the section into `circuit-breaker-proxy` to avoid section ID conflicts

galarragas commented:
@rkuhn : Applied the suggested changes

drewhk commented:
LGTM

patriknw commented:
great contribution, thanks @galarragas
